### ISSUE TITLE
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: Relation fields should never hit the BSON conversion logic.', query-engine/connectors/mongodb-query-connector/src/value.rs:34:35

### ISSUE THREAD

**Issue by ruheni** [prisma/prisma#12929]
### Bug description

Nested write query fails for a 1-n relation – migration from mongoose

### How to reproduce

- Clone repo
- Install dependencies
- Create `docker-compose.yml`:
- Run `docker-compose up -d`
- Seed database
```
git clone git@github.com:prisma/migrate-from-mongoose-to-prisma.git
cd migrate-from-mongoose-to-prisma
npm install
touch docker-compose.yml
touch .env
```

`docker-compose.yml` file
```yml
version: '3'
services:
  mongodb:
   image: 'prismagraphql/mongo-single-replica:4.4.3-bionic'
   environment: 
      MONGO_INITDB_ROOT_USERNAME: root
      MONGO_INITDB_ROOT_PASSWORD: prisma
   ports:
    - 27017:27017
```

Set env variable:
```
DATABASE_URL="mongodb://root:prisma@localhost:27017/prisma-mongo?authSource=admin&retryWrites=true&w=majority"
```

Install Prisma and Prisma Client:
``` 
npm i -D prisma && npm i @prisma/client
```

Initialize Prisma
```
npx prisma init --datasource-provider MongoDB
```

Introspect db:
```
npx prisma db pull
```

Update relations:
```prisma
// schema.prisma

type UsersProfile {
  bio String
}

model Category {
  id      String   @id @default(auto()) @map("_id") @db.ObjectId
  v       Int      @default(0) @map("__v") @ignore
  name    String
  posts   Post[]   @relation(fields: [postIds], references: [id])
  postIds String[] @db.ObjectId

  @@map("categories")
}

model Post {
  id        String  @id @default(auto()) @map("_id") @db.ObjectId
  title     String
  content   String
  published Boolean @default(false)
  v         Int     @default(0) @map("__v") @ignore

  author   User   @relation(fields: [authorId], references: [id])
  authorId String @map("author") @db.ObjectId

  categories  Category[] @relation(fields: [categoryIds], references: [id])
  categoryIds String[]   @map("categories") @db.ObjectId

  @@map("posts")
}

model User {
  id      String        @id @default(auto()) @map("_id") @db.ObjectId
  v       Int           @default(0) @map("__v") @ignore
  email   String        @unique(map: "email_1")
  name    String
  profile UsersProfile?
  posts   Post[]

  @@map("users")
}
```

Push schema

```
npx prisma db push

Update nested write in `src/controller/post.js`

```
import {PrismaClient} from '@prisma/client`
const prisma = new PrismaClient()

const createDraft = async (req, res) => {
  const { title, content, authorEmail } = req.body

  try {
    const draft = await prisma.post.create({
      data: {
        title,
        content,
        author: {
          connect: {
            email: authorEmail
          }
        },
      }
    })

    res.status(201).json(draft)
  } catch (error) {
    return res.status(500).json({ error })
  }
}
```

Start server:
```
RUST_BACKTRACE=1 RUST_BACKTRACE=full npm run dev
```

Make a request to the route:
```
curl -X POST http://localhost:3000/post -H 'Content-Type: application/json' -d '{ "title": "Request", "content": "no funny business", "authorEmail":"nikolas@prisma.io"}'
```

Error message:
```
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: Relation fields should never hit the BSON conversion logic.', query-engine/connectors/mongodb-query-connector/src/value.rs:34:35
stack backtrace:
   0:        0x114c174b1 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4f26ffad025fdbe8
   1:        0x114c3e39b - core::fmt::write::h0a9937d83d3944c1
   2:        0x114c10468 - std::io::Write::write_fmt::hfaf2e2e92eda8127
   3:        0x114c1a0b7 - std::panicking::default_hook::{{closure}}::hc11e9b8d348e68b0
   4:        0x114c19cc5 - std::panicking::default_hook::h1d26ec4d0d63be04
   5:        0x113965c93 - query_engine::engine::set_panic_hook::{{closure}}::hb13d3e90f35b7192
   6:        0x114c1a76e - std::panicking::rust_panic_with_hook::hef4f5e524db188b3
   7:        0x114c1a46e - std::panicking::begin_panic_handler::{{closure}}::h6e8805ea2351af89
   8:        0x114c17927 - std::sys_common::backtrace::__rust_end_short_backtrace::hd383ade987b76f63
   9:        0x114c1a15a - _rust_begin_unwind
  10:        0x114c9b1ef - core::panicking::panic_fmt::hb58956db718d5b79
  11:        0x113d2b77b - mongodb_query_connector::value::<impl mongodb_query_connector::IntoBson for (&prisma_models::field::Field,prisma_value::PrismaValue)>::into_bson::h95e4570bcf56da46
  12:        0x113e1c1cf - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h3e7c9ade7ac96779
  13:        0x113e4cd81 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::ha13e11dd59391f10
  14:        0x113aa7550 - query_core::interpreter::query_interpreters::write::execute::{{closure}}::h8f9f7fb18890375a
  15:        0x113ac886b - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h5c5d66852595493c
  16:        0x113adc8f5 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::heefcba071cfe3997
  17:        0x113ac3d4d - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1e61264390b37814
  18:        0x113ac3d4d - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1e61264390b37814
  19:        0x113ac38cf - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1e61264390b37814
  20:        0x113ad20ee - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h886d3d857f4347e9
  21:        0x1139157cb - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::ha0afe0afe3b0af76
  22:        0x1138f3662 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h0acf6fa6ac03198d
  23:        0x1139044b6 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h449f4b3fca9d1dc3
  24:        0x11392559a - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hda5039d17292f880
  25:        0x113924bea - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hd3090eaeddd40499
  26:        0x1138f755f - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h17bbed97ade6ca82
  27:        0x113907796 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h5567e0f1fccd8808
  28:        0x11392d446 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hf9f999adce7ae502
  29:        0x1138fda4b - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h345089d0b50ec2e5
  30:        0x1138c66b9 - tokio::runtime::task::harness::poll_future::h6184305dc4cac7f4
  31:        0x1138bcdb9 - tokio::runtime::task::harness::Harness<T,S>::poll::h1755f38295a1e100
  32:        0x1149b042f - std::thread::local::LocalKey<T>::with::hf23d5553ac5eca59
  33:        0x1149ac7bb - tokio::runtime::thread_pool::worker::Context::run_task::he6f5c6856859aa9f
  34:        0x1149abb5f - tokio::runtime::thread_pool::worker::Context::run::h41e7f0d9bdf2d560
  35:        0x1149994ba - tokio::macros::scoped_tls::ScopedKey<T>::set::hc3ec82ab4b030874
  36:        0x1149ab654 - tokio::runtime::thread_pool::worker::run::h11ecc88dccbd5de9
  37:        0x114992d43 - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::h27cfa8f24814c06b
  38:        0x1149a384b - tokio::runtime::task::harness::Harness<T,S>::poll::h45a01de20d39e895
  39:        0x114992753 - tokio::runtime::blocking::pool::Inner::run::h4fb0429750936689
  40:        0x11499a501 - std::sys_common::backtrace::__rust_begin_short_backtrace::h4c5fbd74560be1c7
  41:        0x1149b3bed - core::ops::function::FnOnce::call_once{{vtable.shim}}::h85677d5fd49be5ac
  42:        0x114c237b7 - std::sys::unix::thread::Thread::new::thread_start::hcccd55275f17da3c
  43:     0x7ff81a0d64e1 - __pthread_start
```

### Expected behavior

Create a post document and link to existing user

### Prisma information

<!-- Do not include your database credentials when sharing your Prisma schema! -->


### Environment & setup

- OS: Mac OS
- Database: MongoDB
- Node.js version: 16.13.1


### Prisma Version

```
prisma                  : 3.12.0
@prisma/client          : 3.12.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 22b822189f46ef0dc5c5b503
```

**Comment by dpetrick** [comment]
Fix will ship with 3.14.

**Comment by codeinearts** [comment]
Stumbled upon the same error. This still happening  on version `3.15.x` but when doing a regular ´update´ or ´upsert´ on a given model


```
PrismaClientUnknownRequestError: 
Invalid `prisma.zone.upsert()` invocation:


  internal error: entered unreachable code: Relation fields should never hit the BSON conversion logic.
    at RequestHandler.request (/Users/ian.sebastian/Documents/ldev/egenya-main-ui/node_modules/@prisma/client/runtime/index.js:49026:15)
    at async PrismaClient._request (/Users/ian.sebastian/Documents/ldev/egenya-main-ui/node_modules/@prisma/client/runtime/index.js:49919:18) {
  clientVersion: '3.15.1'
}
thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code: Relation fields should never hit the BSON conversion logic.', query-engine/connectors/mongodb-query-connector/src/value.rs:34:35
stack backtrace:
   0:        0x115cd8544 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc80096156cd24709
   1:        0x115cf93fb - core::fmt::write::hb6b947db0e571766
   2:        0x115cd2238 - std::io::Write::write_fmt::h59c4febf5f8fcb1e
   3:        0x115cda22d - std::panicking::default_hook::{{closure}}::h0e76f92bbb367061
   4:        0x115cd9f11 - std::panicking::default_hook::hbbbb8c05ad6cce9a
   5:        0x115cda7ae - std::panicking::rust_panic_with_hook::ha581a2c99399f83b
   6:        0x115cda6a3 - std::panicking::begin_panic_handler::{{closure}}::h97df6e4273b6a879
   7:        0x115cd89c7 - std::sys_common::backtrace::__rust_end_short_backtrace::ha4a3b48ae6bfb343
   8:        0x115cda37a - _rust_begin_unwind
   9:        0x115d654c3 - core::panicking::panic_fmt::h4a5343602e6cbb54
  10:        0x114d500cb - mongodb_query_connector::value::<impl mongodb_query_connector::IntoBson for (&prisma_models::field::Field,prisma_value::PrismaValue)>::into_bson::h79abffa5fb560e82
  11:        0x114d73366 - <query_connector::write_args::ScalarWriteOperation as mongodb_query_connector::root_queries::update::into_operation::IntoUpdateOperation>::into_update_operations::ha40bbdb81f23eaa5
  12:        0x114d77937 - <query_connector::write_args::WriteOperation as mongodb_query_connector::root_queries::update::IntoUpdateDocumentExtension>::into_update_docs::hc3a80b7a65c3ca03
  13:        0x114e67f4f - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hb2aa5eac517a3b4f
  14:        0x114e25bcf - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h160e9f6ddef3bcb4
  15:        0x114c4a33c - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbf548fcbda8118a7
  16:        0x114c45b84 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h8629510d418aaedc
  17:        0x114c45b84 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h8629510d418aaedc
  18:        0x114c4589a - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h8629510d418aaedc
  19:        0x114c479d1 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbe50e5278512c630
  20:        0x114c42753 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h0670ade48d9163b9
  21:        0x114c4c4f8 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbf65784695ad805d
  22:        0x114c45b84 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h8629510d418aaedc
  23:        0x114c4589a - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h8629510d418aaedc
  24:        0x114c479d1 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbe50e5278512c630
  25:        0x1149b0342 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::ha12f86a75681a4b9
  26:        0x114998431 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h53a532419d354381
  27:        0x1149b149f - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hb420c9fe6b4f7cab
  28:        0x114993fd4 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h2dc46389cf006bb5
  29:        0x1149b7020 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hcefdcea7b4374d47
  30:        0x11498ef48 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1ecfbfdf2aec98ea
  31:        0x1149b88f4 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hd518ad129366974b
  32:        0x114994fb3 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h302cd2fc9d6d267f
  33:        0x114958c5d - tokio::runtime::task::harness::poll_future::h2fdd9275ab3d2ed7
  34:        0x11495d7b9 - tokio::runtime::task::harness::Harness<T,S>::poll::h45fa3b791be6114e
  35:        0x1158877e2 - std::thread::local::LocalKey<T>::with::hb86d65c3ab5c1a1f
  36:        0x11587f1ff - tokio::runtime::thread_pool::worker::Context::run_task::h8932529418098e72
  37:        0x11587e5ef - tokio::runtime::thread_pool::worker::Context::run::hac5737482bbc14e3
  38:        0x115868c87 - tokio::macros::scoped_tls::ScopedKey<T>::set::h201fb16f091bc7da
  39:        0x11587dfff - tokio::runtime::thread_pool::worker::run::h77ac0b75b5d8f0be
  40:        0x115863ea3 - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::h17d844e2448304a1
  41:        0x115876f5f - tokio::runtime::task::harness::Harness<T,S>::poll::ha7123904ec8af0ec
  42:        0x115863911 - tokio::runtime::blocking::pool::Inner::run::h1e4cf29d1399a140
  43:        0x11586712f - std::sys_common::backtrace::__rust_begin_short_backtrace::hb1adeafae08b4679
  44:        0x11586bf0d - core::ops::function::FnOnce::call_once{{vtable.shim}}::h8d5ab0e52fb3b63c
  45:        0x115cdf927 - std::sys::unix::thread::Thread::new::thread_start::h1200bb61a12d3334
  46:     0x7ff80cb5f4e1 - __pthread_start
  prisma:client:request_handler  PrismaClientUnknownRequestError: internal error: entered unreachable code: Relation fields should never hit the BSON conversion logic.
    at prismaGraphQLToJSError (/Users/ian.sebastian/Documents/ldev/egenya-main-ui/node_modules/@prisma/client/runtime/index.js:43011:10)
    at LibraryEngine.request (/Users/ian.sebastian/Documents/ldev/egenya-main-ui/node_modules/@prisma/client/runtime/index.js:44865:17)
    at async RequestHandler.request (/Users/ian.sebastian/Documents/ldev/egenya-main-ui/node_modules/@prisma/client/runtime/index.js:48992:24)
    at async PrismaClient._request (/Users/ian.sebastian/Documents/ldev/egenya-main-ui/node_modules/@prisma/client/runtime/index.js:49919:18) {
  clientVersion: '3.15.1'
```

**Comment by codeinearts** [comment]
For future reference, since this is the only place I found something associated with it:

If you define models like this:

```
model Zone {
  parentCluster   Zone?     @relation("ParentChildren", fields: [parentClusterId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  parentClusterId String?   @map("parentCluster") @db.ObjectId   //   <---- here is the problem!
  children        Zone[]    @relation("ParentChildren")
}
```

and try to update model like this:

```
const newZone: Zone = {
  id: data.id,
  name: data.name,
  parentClusterId: data.parentClusterId,
};

const result = await prisma.zone.update({
  where: { id },
  update: {
    parentClusterId: newZone.parentClusterId,
  },
});
```

You may get the error shown, and it explicitly tells you

```
Relation fields should never hit the BSON conversion logic.
```

When modeling, it resulted logic to me to use @map() feature to tell prisma to store `parentClusterId` data under actual `parentCluster` database field (as I draw my model), regardless of relation field being named the same.

Seems like prisma don't like this at all, since also had issues linking data with prisma studio when using this approach on other model fields.

I might be wrong, but couldn't find any place with a mention about this behavior. Removing the @map() attribute makes i work as expected 

````
model Zone {
  parentCluster   Zone?     @relation("ParentChildren", fields: [parentClusterId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  parentClusterId String?   @db.ObjectId   //   <---- this value now sits as parentClusterId on actual MongoDB
  children        Zone[]    @relation("ParentChildren")
}
````

**TL;DR
Do not map your scalar field types using the same field name used on relation field types**

Prisma field definitions:
https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#defining-fields

**Comment by janpio** [comment]
This sounds like unintended behavior @codeinearts - potentially a bug. Can you please open an issue in prisma/prisma so we can reproduce this and hopefully fix it? Thanks. (If we can not fix this, we should not validate a schema that uses field names like this.)

**Comment by peacechen** [comment]
I've also encountered this issue. The error only appears when writing to the table.
```
  company                     Company? @relation(fields: [companyId], references: [id])
  companyId                   String @db.ObjectId @map("company")  // <--- mapping to "company" with the same name as the relation
```

Changing one or the other works:
```
  // Rename company to companyObject
  companyObject               Company? @relation(fields: [companyId], references: [id])
  companyId                   String @db.ObjectId @map("company")
```
